Source/DafnyRuntime/DafnyRuntimeRust/src/lib.rs

```diff
@@ -9,6 +16,346 @@ pub use num::NumCast;
 pub use num::Zero;
 pub use itertools;
 
+// An atomic box is just a RefCell in Rust
+pub type SizeT = usize;
+
+/******************************************************************************
+ * Sequences
+ ******************************************************************************/
+
+// Three elements
+// ArraySequence(var isString: bool, nodeCount: SizeT, immutable_array)
+// ConcatSequence(var isString: bool, nodeCount: SizeT, left: Sequence, right: Sequence)
+// LazySequence(length: BigInteger, box: RefCell<Rc<Sequence<T>>>)
+// We use the named version using {...}, and use snake_case format
+
+// The T must be either a *const T (allocated) OR a Reference Counting (immutable)
+
+#[allow(dead_code)]
+mod dafny_runtime_conversions {
+    use num::BigInt;
+
+    use crate::Sequence;
+    use crate::Map;
+    use std::cell::RefCell;
+    use std::collections::HashMap;
+    use std::rc::Rc;
+    use std::hash::Hash;
+    pub fn dafny_int_to_bigint(i: &Rc<BigInt>) -> BigInt {
+        i.as_ref().clone()
+    }
+    pub fn bigint_to_dafny_int(i: &BigInt) -> Rc<BigInt> {
+        Rc::new(i.clone())
+    }
+
+    pub fn dafny_sequence_to_vec<T, X>(s: &Sequence<T>, elem_converter: fn(&T) -> X) -> Vec<X>
+      where T: Clone
+    {
+        let mut array: Vec<T> = Vec::with_capacity(s.cardinality());
+        Sequence::<T>::append_recursive(&mut array, s);
+        array.iter().map(|x| elem_converter(x)).collect()
+    }
+
+    // Used for external conversions
+    pub fn vec_to_dafny_sequence<T, X>(array: &Vec<X>, elem_converter: fn(&X) -> T) -> Rc<Sequence<T>>
+      where T: Clone
+    {
+        let mut result: Vec<T> = Vec::with_capacity(array.len());
+        for elem in array.iter() {
+            result.push(elem_converter(elem));
+        }
+        Sequence::<T>::new_array_sequence(&Rc::new(result))
+    }
+    
+    fn dafny_map_to_hashmap<K, V, K2, V2>(m: &Rc<Map<K, V>>, converter_k: fn(&K)->K2, converter_v: fn(&V)->V2) -> HashMap<K2, V2>
+      where
+          K: Eq + Clone + Hash, V: Clone,
+          K2: Eq + Hash, V2: Clone
+    {
+        m.to_hashmap_owned(converter_k, converter_v)
+    }
+
+    fn hashmap_to_dafny_map<K2, V2, K, V>(map: &HashMap<K2, V2>, converter_k: fn(&K2)->K, converter_v: fn(&V2)->V)
+        -> Rc<Map<K, V>>
+      where
+        K: Eq + Clone + Hash, V: Clone,
+        K2: Eq + Hash, V2: Clone
+    {
+        let mut result: Vec<(K, V)> = Vec::new();
+        for (k, v) in map.iter() {
+            result.push((converter_k(k), converter_v(v)));
+        }
+        let s = Sequence::<(K, V)>::new_array_sequence(&Rc::new(result));
+        Rc::new(Map {
+            data: s,
+            cache: RefCell::new(None)
+        })
+    }
+
+    // --unicode-chars:true
+    pub mod unicode_chars_true {
+        use crate::Sequence;
+        use std::rc::Rc;
+
+        pub fn string_to_dafny_string(s: &str) -> Rc<Sequence<char>> {
```

Comment:
Definitely try to introduce those aliases I mentioned next PR if possible, `dafny_string` instead of `Rc<Sequence<char>>`

Reply:
I introduced them two weeks ago in the upcoming work, definitely a good idea.